### ISSUE TITLE
trtllm_fp4_block_scale_routed_moe has larger error for small number of tokens

### ISSUE THREAD

**Issue by sshleifer** [flashinfer-ai/flashinfer#3559]
`fused_moe.trtllm_fp4_block_scale_routed_moe` produces batch-size-dependent
output at small batch: the same token — byte-identical hidden_states/scales/weights,
pre-routed topk_ids — gets a different result depending on how many other tokens are
in the batch (and the result is further from fp32 at bsz=8 than if we pad to bsz=64). Wondering if this is fixable without padding.

### Environment

-  flashinfer 0.6.8.post1 (+ GPU arch — these repros are on sm_100/GB300)

### Repro and Results

Same 8 tokens, identical static scales/weights, only the surrounding batch differs:

| run | p99 \|err\| vs **fp32** torch ref | max \|err\| | (out range ~4) |
|---|---|---|---|
| tokens 0–7 as a **T=8** batch | 0.33 | 0.80 | degraded |
| tokens 0–7 embedded in a **T=64** batch | 0.05 | 0.12 | clean fp4 floor |

```python
"""Self-contained repro: trtllm_fp4_block_scale_routed_moe is batch-size-dependent.
Same tokens + same static scales/weights give different output at T=8 vs embedded in T=64."""
import torch
from flashinfer import RoutingMethodType, fp4_quantize, nvfp4_block_scale_interleave
from flashinfer.fused_moe import trtllm_fp4_block_scale_routed_moe
from flashinfer.fused_moe.core import (
    _maybe_get_cached_w3_w1_permute_indices, get_w2_permute_indices_with_cache)

dev = "cuda"; torch.manual_seed(0)
FP8 = torch.float8_e4m3fn; FP8_MAX, FP4_MAX = 448.0, 6.0
E, D, F, top_k, Tbig = 256, 4096, 512, 6, 64

def gscale(t): return (FP8_MAX * FP4_MAX / t.abs().max().clamp(min=1e-12)).float()

# --- weights: quantize logical bf16 -> fp4, lay out for trtllm (all tensors on cuda) ---
w13 = torch.randn(E, 2*F, D, device=dev, dtype=torch.bfloat16) / 10
w2  = torch.randn(E, D, F, device=dev, dtype=torch.bfloat16) / 10
w13_gs = torch.stack([gscale(w13[e]) for e in range(E)])
w2_gs  = torch.stack([gscale(w2[e])  for e in range(E)])
def prep13():
    c={};P=[];S=[]
    for e in range(E):
        p,s = fp4_quantize(w13[e], w13_gs[e], 16, False, False)
        p=p.reshape(2*F, D//2); s=s.view(FP8).reshape(2*F, D//16).view(torch.uint8)
        pi=_maybe_get_cached_w3_w1_permute_indices(c,p,128,is_gated_act_gemm=True); P.append(p[pi.to(dev)].contiguous())
        ps=_maybe_get_cached_w3_w1_permute_indices(c,s,128,num_elts_per_sf=16,is_gated_act_gemm=True); S.append(nvfp4_block_scale_interleave(s[ps.to(dev)].contiguous()))
    return torch.stack(P), torch.stack(S).view(FP8).reshape(E,2*F,D//16)
def prep2():
    c={};P=[];S=[]
    for e in range(E):
        p,s = fp4_quantize(w2[e], w2_gs[e], 16, False, False)
        p=p.reshape(D, F//2); s=s.view(FP8).reshape(D, F//16).view(torch.uint8)
        pi=get_w2_permute_indices_with_cache(c,p,128); P.append(p[pi.to(dev)].contiguous())
        ps=get_w2_permute_indices_with_cache(c,s,128,num_elts_per_sf=16); S.append(nvfp4_block_scale_interleave(s[ps.to(dev)].contiguous()))
    return torch.stack(P), torch.stack(S).view(FP8).reshape(E,D,F//16)
W1,W1s = prep13(); W2,W2s = prep2()

# --- activations + routing (cuda); a2 global scale ~ intermediate amax ---
x  = torch.randn(Tbig, D, device=dev, dtype=torch.bfloat16) / 10
a_gs = gscale(x)
xf, xs = fp4_quantize(x, a_gs, 16, False, False)
xf = xf.reshape(Tbig, D//2); xs = xs.view(FP8).reshape(Tbig, D//16)
ids = torch.randint(0, E, (Tbig, top_k), dtype=torch.int32, device=dev)
wts = torch.rand(Tbig, top_k, device=dev, dtype=torch.bfloat16)
a2 = torch.tensor(FP8_MAX*FP4_MAX/8.0, device=dev)
g1_alphas = ((1.0/w13_gs)/a_gs).float(); g1_scale_c = (g1_alphas*a2).float(); g2_alphas = ((1.0/w2_gs)/a2).float()

def run(T):
    p = (ids[:T] << 16) | wts[:T].view(torch.int16).to(torch.int32)
    return trtllm_fp4_block_scale_routed_moe(
        p, None, xf[:T], xs[:T], W1, W1s, None, None, None, None, W2, W2s, None,
        g1_scale_c, g1_alphas, g2_alphas, num_experts=E, top_k=top_k, n_group=None,
        topk_group=None, intermediate_size=F, local_expert_offset=0, local_num_experts=E,
        routed_scaling_factor=None, routing_method_type=RoutingMethodType.TopK,
        activation_type=3, tune_max_num_tokens=Tbig)[0]

o8, o64 = run(8), run(Tbig)
print(f"same 8 tokens, T=8 vs embedded in T=64: max|diff| = {(o8.float()-o64[:8].float()).abs().max():.4f}")
print(f"(output range [{o64.min():.2f},{o64.max():.2f}]; expect ~0, observe nonzero = the bug)")
```

**Comment by sshleifer** [comment]
- not every routing triggers the bug

**Comment by sshleifer** [comment]
bad setup. Routing had duplicates. Sorry!
